man/mysql.1
'\" t
.\"     Title: mysql
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
<<<<<<< HEAD
.\"      Date: 11/26/2022
||||||| merged common ancestors
<<<<<<<<< Temporary merge branch 1
.\"      Date: 12/07/2022
||||||||| merged common ancestors
.\"      Date: 08/29/2022
=========
.\"      Date: 11/26/2022
>>>>>>>>> Temporary merge branch 2
=======
.\"      Date: 03/03/2023
>>>>>>> Percona-Server-8.0.33-25
.\"    Manual: MySQL Database System
.\"    Source: MySQL 8.0
.\"  Language: English
.\"
<<<<<<< HEAD
.TH "MYSQL" "1" "11/26/2022" "MySQL 8\&.0" "MySQL Database System"
||||||| merged common ancestors
<<<<<<<<< Temporary merge branch 1
.TH "MYSQL" "1" "12/07/2022" "MySQL 5\&.7" "MySQL Database System"
||||||||| merged common ancestors
.TH "MYSQL" "1" "08/29/2022" "MySQL 5\&.7" "MySQL Database System"
=========
.TH "MYSQL" "1" "11/26/2022" "MySQL 8\&.0" "MySQL Database System"
>>>>>>>>> Temporary merge branch 2
=======
.TH "MYSQL" "1" "03/03/2023" "MySQL 8\&.0" "MySQL Database System"
>>>>>>> Percona-Server-8.0.33-25
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mysql \- the MySQL command\-line client
.SH "SYNOPSIS"
.HP \w'\fBmysql\ [\fR\fB\fIoptions\fR\fR\fB]\ \fR\fB\fIdb_name\fR\fR\ 'u
\fBmysql [\fR\fB\fIoptions\fR\fR\fB] \fR\fB\fIdb_name\fR\fR
.SH "DESCRIPTION"
.PP
\fBmysql\fR
is a simple SQL shell with input line editing capabilities\&. It supports interactive and noninteractive use\&. When used interactively, query results are presented in an ASCII\-table format\&. When used noninteractively (for example, as a filter), the result is presented in tab\-separated format\&. The output format can be changed using command options\&.
.PP
If you have problems due to insufficient memory for large result sets, use the
\fB\-\-quick\fR
option\&. This forces
\fBmysql\fR
to retrieve results from the server a row at a time rather than retrieving the entire result set and buffering it in memory before displaying it\&. This is done by returning the result set using the
mysql_use_result()
C API function in the client/server library rather than
mysql_store_result()\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
Alternatively, MySQL Shell offers access to the X DevAPI\&. For details, see
\m[blue]\fBMySQL Shell 8\&.0\fR\m[]\&\s-2\u[1]\d\s+2\&.
.sp .5v
.RE
.PP
Using
\fBmysql\fR
is very easy\&. Invoke it from the prompt of your command interpreter as follows:
.sp
.if n \{\
.RS 4
.\}
.nf
mysql \fIdb_name\fR
.fi
.if n \{\
.RE
.\}
.PP
Or:
.sp
.if n \{\
.RS 4
.\}
.nf
mysql \-\-user=\fIuser_name\fR \-\-password \fIdb_name\fR
.fi
.if n \{\
.RE
.\}
.PP
In this case, you\*(Aqll need to enter your password in response to the prompt that
\fBmysql\fR
displays:
.sp
.if n \{\
.RS 4
.\}
.nf
Enter password: \fIyour_password\fR
.fi
.if n \{\
.RE
.\}
.PP
Then type an SQL statement, end it with
;,
\eg, or
\eG
and press Enter\&.
.PP
Typing
Control+C
interrupts the current statement if there is one, or cancels any partial input line otherwise\&.
.PP
You can execute SQL statements in a script file (batch file) like this:
.sp
.if n \{\
.RS 4
.\}
.nf
mysql \fIdb_name\fR < \fIscript\&.sql\fR > \fIoutput\&.tab\fR
.fi
.if n \{\
.RE
.\}
.PP
On Unix, the
\fBmysql\fR
client logs statements executed interactively to a history file\&. See
the section called \(lqMYSQL CLIENT LOGGING\(rq\&.
.SH "MYSQL CLIENT OPTIONS"
.PP
\fBmysql\fR
supports the following options, which can be specified on the command line or in the
[mysql]
and
[client]
groups of an option file\&. For information about option files used by MySQL programs, see
Section\ \&4.2.2.2, \(lqUsing Option Files\(rq\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-help\fR,
\fB\-?\fR
Display a help message and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-auto\-rehash\fR
Enable automatic rehashing\&. This option is on by default, which enables database, table, and column name completion\&. Use
\fB\-\-disable\-auto\-rehash\fR
to disable rehashing\&. That causes
\fBmysql\fR
to start faster, but you must issue the
rehash
command or its
\e#
shortcut if you want to use name completion\&.
.sp
To complete a name, enter the first part and press Tab\&. If the name is unambiguous,
\fBmysql\fR
completes it\&. Otherwise, you can press Tab again to see the possible names that begin with what you have typed so far\&. Completion does not occur if there is no default database\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
This feature requires a MySQL client that is compiled with the
\fBreadline\fR
library\&. Typically, the
\fBreadline\fR
library is not available on Windows\&.
.sp .5v
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-auto\-vertical\-output\fR
Cause result sets to be displayed vertically if they are too wide for the current window, and using normal tabular format otherwise\&. (This applies to statements terminated by
;
or
\eG\&.)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-batch\fR,
\fB\-B\fR
Print results using tab as the column separator, with each row on a new line\&. With this option,
\fBmysql\fR
does not use the history file\&.
.sp
Batch mode results in nontabular output format and escaping of special characters\&. Escaping may be disabled by using raw mode; see the description for the
\fB\-\-raw\fR
option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-binary\-as\-hex\fR
When this option is given,
\fBmysql\fR
displays binary data using hexadecimal notation (0x\fIvalue\fR)\&. This occurs whether the overall output display format is tabular, vertical, HTML, or XML\&.
.sp
\fB\-\-binary\-as\-hex\fR
when enabled affects display of all binary strings, including those returned by functions such as
CHAR()
and
UNHEX()\&. The following example demonstrates this using the ASCII code for
A
(65 decimal, 41 hexadecimal):
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-binary\-as\-hex\fR
disabled:
.sp
.if n \{\
.RS 4
.\}
.nf
mysql> \fBSELECT CHAR(0x41), UNHEX(\*(Aq41\*(Aq);\fR
+\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-+
| CHAR(0x41) | UNHEX(\*(Aq41\*(Aq) |
+\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-+
| A          | A           |
+\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-+
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-binary\-as\-hex\fR
enabled:
.sp
.if n \{\
.RS 4
.\}
.nf
mysql> \fBSELECT CHAR(0x41), UNHEX(\*(Aq41\*(Aq);\fR
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
| CHAR(0x41)             | UNHEX(\*(Aq41\*(Aq)              |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
| 0x41                   | 0x41                     |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
.fi
.if n \{\
.RE
.\}
.RE
.sp
To write a binary string expression so that it displays as a character string regardless of whether
\fB\-\-binary\-as\-hex\fR
is enabled, use these techniques:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The
CHAR()
function has a
USING \fIcharset\fR
clause:
.sp
.if n \{\
.RS 4
.\}
.nf
mysql> \fBSELECT CHAR(0x41 USING utf8mb4);\fR
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
| CHAR(0x41 USING utf8mb4) |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
| A                        |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
More generally, use
CONVERT()
to convert an expression to a given character set:
.sp
.if n \{\
.RS 4
.\}
.nf
mysql> \fBSELECT CONVERT(UNHEX(\*(Aq41\*(Aq) USING utf8mb4);\fR
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
| CONVERT(UNHEX(\*(Aq41\*(Aq) USING utf8mb4) |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
| A                                  |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
.fi
.if n \{\
.RE
.\}
.RE
.sp
As of MySQL 8\&.0\&.19, when
\fBmysql\fR
operates in interactive mode, this option is enabled by default\&. In addition, output from the
status
(or
\es) command includes this line when the option is enabled implicitly or explicitly:
.sp
.if n \{\
.RS 4
.\}
.nf
Binary data as: Hexadecimal
.fi
.if n \{\
.RE
.\}
.sp
To disable hexadecimal notation, use
\fB\-\-skip\-binary\-as\-hex\fR
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-binary\-mode\fR
This option helps when processing
\fBmysqlbinlog\fR
output that may contain
BLOB
values\&. By default,
\fBmysql\fR
translates
\er\en
in statement strings to
\en
and interprets
\e0
as the statement terminator\&.
\fB\-\-binary\-mode\fR
disables both features\&. It also disables all
\fBmysql\fR
commands except
charset
and
delimiter
in noninteractive mode (for input piped to
\fBmysql\fR
or loaded using the
source
command)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-bind\-address=\fR\fB\fIip_address\fR\fR
On a computer having multiple network interfaces, use this option to select which interface to use for connecting to the MySQL server\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-character\-sets\-dir=\fR\fB\fIdir_name\fR\fR
The directory where character sets are installed\&. See
Section\ \&10.15, \(lqCharacter Set Configuration\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-column\-names\fR
Write column names in results\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-column\-type\-info\fR
Display result set metadata\&. This information corresponds to the contents of C API
MYSQL_FIELD
data structures\&. See
\m[blue]\fBC API Basic Data Structures\fR\m[]\&\s-2\u[2]\d\s+2\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-comments\fR,
\fB\-c\fR
Whether to strip or preserve comments in statements sent to the server\&. The default is
\fB\-\-skip\-comments\fR
(strip comments), enable with
\fB\-\-comments\fR
(preserve comments)\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
The
\fBmysql\fR
client always passes optimizer hints to the server, regardless of whether this option is given\&.
.sp
Comment stripping is deprecated\&. Expect this feature and the options to control it to be removed in a future MySQL release\&.
.sp .5v
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-compress\fR,
\fB\-C\fR
Compress all information sent between the client and the server if possible\&. See
Section\ \&4.2.8, \(lqConnection Compression Control\(rq\&.
.sp
As of MySQL 8\&.0\&.18, this option is deprecated\&. Expect it be removed in a future version of MySQL\&. See
the section called \(lqConfiguring Legacy Connection Compression\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-compression\-algorithms=\fR\fB\fIvalue\fR\fR
The permitted compression algorithms for connections to the server\&. The available algorithms are the same as for the
protocol_compression_algorithms
system variable\&. The default value is
uncompressed\&.
.sp
For more information, see
Section\ \&4.2.8, \(lqConnection Compression Control\(rq\&.
.sp
This option was added in MySQL 8\&.0\&.18\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-connect\-expired\-password\fR
Indicate to the server that the client can handle sandbox mode if the account used to connect has an expired password\&. This can be useful for noninteractive invocations of
\fBmysql\fR
because normally the server disconnects noninteractive clients that attempt to connect using an account with an expired password\&. (See
Section\ \&6.2.16, \(lqServer Handling of Expired Passwords\(rq\&.)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-connect\-timeout=\fR\fB\fIvalue\fR\fR
The number of seconds before connection timeout\&. (Default value is
0\&.)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-database=\fR\fB\fIdb_name\fR\fR,
\fB\-D \fR\fB\fIdb_name\fR\fR
The database to use\&. This is useful primarily in an option file\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-debug[=\fR\fB\fIdebug_options\fR\fR\fB]\fR,
\fB\-# [\fR\fB\fIdebug_options\fR\fR\fB]\fR
Write a debugging log\&. A typical
\fIdebug_options\fR
string is
d:t:o,\fIfile_name\fR\&. The default is
d:t:o,/tmp/mysql\&.trace\&.
.sp
This option is available only if MySQL was built using
\fBWITH_DEBUG\fR\&. MySQL release binaries provided by Oracle are
\fInot\fR
built using this option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-debug\-check\fR
Print some debugging information when the program exits\&.
.sp
This option is available only if MySQL was built using
\fBWITH_DEBUG\fR\&. MySQL release binaries provided by Oracle are
\fInot\fR
built using this option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}

Docs/ChangeLog
Placeholder



support-files/mysql.server.sh
#!/bin/sh
# Copyright Abandoned 1996 TCX DataKonsult AB & Monty Program KB & Detron HB
# This file is public domain and comes with NO WARRANTY of any kind

# MySQL (Percona XtraDB Cluster) daemon start/stop script.

# Usually this is put in /etc/init.d (at least on machines SYSV R4 based
# systems) and linked to /etc/rc3.d/S99mysql and /etc/rc0.d/K01mysql.
# When this is done the mysql server will be started when the machine is
# started and shut down when the systems goes down.

# Comments to support chkconfig on RedHat Linux
# chkconfig: 2345 64 36
# description: A very fast and reliable SQL database engine.

# Comments to support LSB init script conventions
### BEGIN INIT INFO
# Provides: mysql
# Required-Start: $local_fs $network $remote_fs
# Should-Start: ypbind nscd ldap ntpd xntpd
# Required-Stop: $local_fs $network $remote_fs
# Default-Start:  2 3 4 5
# Default-Stop: 0 1 6
# Short-Description: start and stop MySQL (Percona XtraDB Cluster)
# Description: Percona-Server is a SQL database engine with focus on high performance.
### END INIT INFO
 
# Prevent OpenSUSE's init scripts from calling systemd, so that
# both 'bootstrap' and 'start' are handled entirely within this script

SYSTEMD_NO_WRAP=1

# Prevent Debian's init scripts from calling systemctl

SYSTEMCTL_SKIP_REDIRECT=true
_SYSTEMCTL_SKIP_REDIRECT=true
 
# If you install MySQL on some other places than @prefix@, then you
# have to do one of the following things for this script to work:
#
# - Run this script from within the MySQL installation directory
# - Create a /etc/my.cnf file with the following information:
#   [mysqld]
#   basedir=<path-to-mysql-installation-directory>
# - Add the above to any other configuration file (for example ~/.my.ini)
#   and copy my_print_defaults to /usr/bin
# - Add the path to the mysql-installation-directory to the basedir variable
#   below.
#
# If you want to affect other MySQL variables, you should make your changes
# in the /etc/my.cnf, ~/.my.cnf or other MySQL configuration files.

# If you change base dir, you must also change datadir. These may get
# overwritten by settings in the MySQL configuration files.

# source the defauts file /etc/sysconfig/mysql if it exists for any environment variables.
# set -a is done so that  'FOO=BAR' just be used than 'export FOO=BAR' in the /etc/sysconfig/mysql.
# However, export FOO=BAR in that file also should work.
set -a
[ -r /etc/sysconfig/mysql ] && . /etc/sysconfig/mysql
set +a

basedir=
datadir=

# Default value, in seconds, afterwhich the script should timeout waiting
# for server start. 
# Value here is overriden by value in my.cnf. 
# 0 means don't wait at all
# Negative numbers mean to wait indefinitely
service_startup_timeout=900
startup_sleep=1

[ -e /etc/sysconfig/mysql ] && . /etc/sysconfig/mysql

# Lock directory for RedHat / SuSE.
lockdir='/var/lock/subsys'
lock_file_path="$lockdir/mysql"

# The following variables are only set for letting mysql.server find things.

# Set some defaults
mysqld_pid_file_path=
if test -z "$basedir"
then
  basedir=@prefix@
  bindir=@bindir@
  if test -z "$datadir"
  then
    datadir=@localstatedir@
  fi
  sbindir=@sbindir@
  libexecdir=@libexecdir@
else
  bindir="$basedir/bin"
  if test -z "$datadir"
  then
    datadir="$basedir/data"
  fi
  sbindir="$basedir/sbin"
  libexecdir="$basedir/libexec"
fi

# datadir_set is used to determine if datadir was set (and so should be
# *not* set inside of the --basedir= handler.)
datadir_set=

#
# Use LSB init script functions for printing messages, if possible
#
lsb_functions="/lib/lsb/init-functions"
if test -f $lsb_functions ; then
  . $lsb_functions
else
  log_success_msg()
  {
    echo " SUCCESS! $@"
  }
  log_failure_msg()
  {
    echo " ERROR! $@"
  }
fi

PATH="/sbin:/usr/sbin:/bin:/usr/bin:$basedir/bin"
export PATH

mode=$1    # start or stop

[ $# -ge 1 ] && shift


other_args="$*"   # uncommon, but needed when called from an RPM upgrade action
           # Expected: "--skip-networking --skip-grant-tables"
           # They are not checked here, intentionally, as it is the resposibility
           # of the "spec" file author to give correct arguments only.

case `echo "testing\c"`,`echo -n testing` in
    *c*,-n*) echo_n=   echo_c=     ;;
    *c*,*)   echo_n=-n echo_c=     ;;
    *)       echo_n=   echo_c='\c' ;;
esac

parse_server_arguments() {
  for arg do
    case "$arg" in
      --basedir=*)  basedir=`echo "$arg" | sed -e 's/^[^=]*=//'`
                    bindir="$basedir/bin"
		    if test -z "$datadir_set"; then
		      datadir="$basedir/data"
		    fi
		    sbindir="$basedir/sbin"
		    libexecdir="$basedir/libexec"
        ;;
      --datadir=*)  datadir=`echo "$arg" | sed -e 's/^[^=]*=//'`
		    datadir_set=1
	;;
      --pid-file=*|--pid_file=*) mysqld_pid_file_path=`echo "$arg" | sed -e 's/^[^=]*=//'` ;;
      --service-startup-timeout=*|--service_startup_timeout=*) service_startup_timeout=`echo "$arg" | sed -e 's/^[^=]*=//'` ;;
    esac
  done
}

wait_for_pid () {
  verb="$1"           # created | removed
  pid="$2"            # process ID of the program operating on the pid-file
  pid_file_path="$3" # path to the PID file.

  sst_progress_file=$datadir/sst_in_progress
  i=0
  avoid_race_condition="by checking again"

  while test $i -ne $service_startup_timeout ; do

    case "$verb" in
      'created')
        # wait for a PID-file to pop into existence.
        test -s "$pid_file_path" && i='' && break
        ;;
      'removed')
        # wait for this PID-file to disappear
        test ! -s "$pid_file_path" && i='' && break
        ;;
      *)
        echo "wait_for_pid () usage: wait_for_pid created|removed pid pid_file_path"
        exit 1
        ;;
    esac

    # if server isn't running, then pid-file will never be updated
    if test -n "$pid"; then
      if kill -0 "$pid" 2>/dev/null; then
        :  # the server still runs
      else
        # The server may have exited between the last pid-file check and now.  
        if test -n "$avoid_race_condition"; then
          avoid_race_condition=""
          continue  # Check again.
        fi

        # there's nothing that will affect the file.
        log_failure_msg "The server quit without updating PID file ($pid_file_path)."
        return 1  # not waiting any more.
      fi
    fi

    if (test -e $sst_progress_file || [ $mode = 'bootstrap-pxc' ]) && [ $startup_sleep -ne 10 ];then
        echo "State transfer in progress, setting sleep higher"
        startup_sleep=10
    fi

    echo $echo_n ".$echo_c"
    i=`expr $i + 1`
    sleep $startup_sleep

  done

  if test -z "$i" ; then
    log_success_msg
    return 0
  elif test -e $sst_progress_file; then 
    log_failure_msg
    return 2
  else
    log_failure_msg
    return 1
  fi
}

# Get arguments from the my.cnf file,
# the only group, which is read from now on is [mysqld]
if test -x ./bin/my_print_defaults
then
  print_defaults="./bin/my_print_defaults"
elif test -x $bindir/my_print_defaults
then
  print_defaults="$bindir/my_print_defaults"
elif test -x $bindir/mysql_print_defaults
then
  print_defaults="$bindir/mysql_print_defaults"
else
  # Try to find basedir in /etc/my.cnf
  conf=/etc/my.cnf
  print_defaults=
  if test -r $conf
  then
    subpat='^[^=]*basedir[^=]*=\(.*\)$'
    dirs=`sed -e "/$subpat/!d" -e 's//\1/' $conf`
    for d in $dirs
    do
      d=`echo $d | sed -e 's/[ 	]//g'`
      if test -x "$d/bin/my_print_defaults"
      then
        print_defaults="$d/bin/my_print_defaults"
        break
      fi
      if test -x "$d/bin/mysql_print_defaults"
      then
        print_defaults="$d/bin/mysql_print_defaults"
        break
      fi
    done
  fi

  # Hope it's in the PATH ... but I doubt it
  test -z "$print_defaults" && print_defaults="my_print_defaults"
fi

#
# Read defaults file from 'basedir'.   If there is no defaults file there
# check if it's in the old (depricated) place (datadir) and read it from there
#

extra_args=""
if test -r "$basedir/my.cnf"
then
  extra_args="-e $basedir/my.cnf"
else
  if test -r "$datadir/my.cnf"
  then
    extra_args="-e $datadir/my.cnf"
  fi
fi

parse_server_arguments `$print_defaults $extra_args mysqld server mysql_server mysql.server`

#
# Set pid file if not given
#
if test -z "$mysqld_pid_file_path"
then
  mysqld_pid_file_path=$datadir/`@HOSTNAME@`.pid
else
  case "$mysqld_pid_file_path" in
    /* ) ;;
    * )  mysqld_pid_file_path="$datadir/$mysqld_pid_file_path" ;;
  esac
fi

check_running() {

    local show_msg=$1

    # First, check to see if pid file exists
    if test -s "$mysqld_pid_file_path" ; then 
        read mysqld_pid < "$mysqld_pid_file_path"
        if kill -0 $mysqld_pid 2>/dev/null ; then 
            log_success_msg "MySQL (Percona XtraDB Cluster) running ($mysqld_pid)"
            return 0
        else
            log_failure_msg "MySQL (Percona XtraDB Cluster) is not running, but PID file exists"
            return 1
        fi
    else
        # Try to find appropriate mysqld process
        mysqld_pid=`pidof $libexecdir/mysqld`

        # test if multiple pids exist
        pid_count=`echo $mysqld_pid | wc -w`
        if test $pid_count -gt 1 ; then
            log_failure_msg "Multiple MySQL running but PID file could not be found ($mysqld_pid)"
            return 5
        elif test -z $mysqld_pid ; then 
            if test -f "$lock_file_path" ; then 
                log_failure_msg "MySQL (Percona XtraDB Cluster) is not running, but lock file ($lock_file_path) exists"
                return 3
            fi 
            test $show_msg -eq 1 && log_failure_msg "MySQL (Percona XtraDB Cluster) is not running"
            return 3
        else
            log_failure_msg "MySQL (Percona XtraDB Cluster) is running but PID file could not be found"
            return 4
        fi
    fi
}

case "$mode" in
  'start')

    check_running 0
    ext_status=$?
    if test $ext_status -ne 3;then 
        exit $ext_status
    fi

    # Start daemon
    if test -e $datadir/sst_in_progress;then 
        echo "Stale sst_in_progress file in datadir"
    fi

    # Safeguard (relative paths, core dumps..)
    cd $basedir

    echo $echo_n "Starting MySQL (Percona XtraDB Cluster)"
    if test -x $bindir/mysqld_safe
    then
      # Give extra arguments to mysqld with the my.cnf file. This script
      # may be overwritten at next upgrade.
      $bindir/mysqld_safe --datadir="$datadir" --pid-file="$mysqld_pid_file_path" $other_args >/dev/null &
      wait_for_pid created "$!" "$mysqld_pid_file_path"; return_value=$?
      if [ $return_value == 1 ];then 
          log_failure_msg "MySQL (Percona XtraDB Cluster) server startup failed!"
      elif [ $return_value == 2 ];then
          log_failure_msg "MySQL (Percona XtraDB Cluster) server startup failed! State transfer still in progress"
      fi

      # Make lock for RedHat / SuSE
      if test -w "$lockdir"
      then
        touch "$lock_file_path"
      fi

      exit $return_value
    else
      log_failure_msg "Couldn't find MySQL server ($bindir/mysqld_safe)"
    fi
    ;;

  'stop')
    # Stop daemon. We use a signal here to avoid having to know the
    # root password.
    echo $echo_n "Shutting down MySQL (Percona XtraDB Cluster)"
    if test -s "$mysqld_pid_file_path"
    then
      mysqld_pid=`cat "$mysqld_pid_file_path"`

      if (kill -0 $mysqld_pid 2>/dev/null)
      then
        kill $mysqld_pid
        # mysqld should remove the pid file when it exits, so wait for it.
        wait_for_pid removed "$mysqld_pid" "$mysqld_pid_file_path"; return_value=$?
        if [ $return_value != 0 ];then 
            log_failure_msg "MySQL (Percona XtraDB Cluster) server stop failed!"
        fi
      else
        log_failure_msg "MySQL (Percona XtraDB Cluster) server process #$mysqld_pid is not running!"
        rm "$mysqld_pid_file_path"
      fi

      # To avoid the race condition involved in restarts with PID file.
      if (kill -0 $mysqld_pid 2>/dev/null)
      then
          sleep 2
      fi

      # Delete lock for RedHat / SuSE
      if test -f "$lock_file_path"
      then
        rm -f "$lock_file_path"
      fi
      exit $return_value
    else
      log_failure_msg "MySQL (Percona XtraDB Cluster) PID file could not be found!"
    fi
    ;;

  'restart')
    # Stop the service and regardless of whether it was
    # running or not, start it again.
    if $0 stop  $other_args; then
      if ! $0 start $other_args; then 
         log_failure_msg "Failed to restart server."
         exit 1
      fi
    else
      log_failure_msg "Failed to stop running server, so refusing to try to start."
      exit 1
    fi
    ;;

  'restart-bootstrap')
    # Stop the service and regardless of whether it was
    # running or not, start it again.
    if $0 stop  $other_args; then
      if ! $0 bootstrap-pxc $other_args;then
         log_failure_msg "Failed to restart server with bootstrap."
         exit 1
      fi
    else
      log_failure_msg "Failed to stop running server, so refusing to try to start."
      exit 1
    fi
    ;;

  'reload'|'force-reload')
    if test -s "$mysqld_pid_file_path" ; then
      read mysqld_pid <  "$mysqld_pid_file_path"
      kill -HUP $mysqld_pid && log_success_msg "Reloading service MySQL (Percona XtraDB Cluster)"
      touch "$mysqld_pid_file_path"
    else
      log_failure_msg "MySQL (Percona XtraDB Cluster) PID file could not be found!"
      exit 1
    fi
    ;;
  'status')
      check_running 1
      exit $?
    ;;
  'bootstrap-pxc')
      # Bootstrap the Percona XtraDB Cluster, start the first node
      # that initiate the cluster
      echo $echo_n "Bootstrapping PXC (Percona XtraDB Cluster)"
      $0 start $other_args --wsrep-new-cluster
      ;;
    'bootstrap')
      # Bootstrap the cluster, start the first node
      # that initiate the cluster
      echo $echo_n "Bootstrapping the cluster"
      $0 start $other_args --wsrep-new-cluster
      ;;
    *)
      # usage
      basename=`basename "$0"`
      echo "Usage: $basename {start|stop|restart|restart-bootstrap|reload|force-reload|status|bootstrap-pxc}  [ MySQL (Percona XtraDB Cluster) options ]"
      exit 1
    ;;
esac

exit 0
